internal/extender/resource.go

```diff
@@ -415,6 +415,75 @@ func (s *SparkSchedulerExtender) getNodes(ctx context.Context, nodeNames []strin
 	return availableNodes
 }
 
+func filterNodesToZone(ctx context.Context, initialNodes []*v1.Node, zone string) ([]*v1.Node, error) {
+	nodes := make([]*v1.Node, 0, len(initialNodes))
+	for _, node := range initialNodes {
+		zoneLabel, ok := node.Labels[v1.LabelTopologyZone]
+		if !ok {
+			return nil, werror.ErrorWithContextParams(ctx, "Could not read zone label from node, unable to make scheduling decisions based on AZ")
+		}
+
+		if zoneLabel == zone {
+			nodes = append(nodes, node)
+		}
+	}
+	svc1log.FromContext(ctx).Info("Filtered nodes in zone", svc1log.SafeParam("zone", zone), svc1log.SafeParam("nodes", nodes))
+	return nodes, nil
+}
+
+func (s *SparkSchedulerExtender) getCommonZoneForExecutorsApplication(ctx context.Context, executor *v1.Pod) (string, error) {
+	executorSparkLabel, ok := executor.Labels[common.SparkAppIDLabel]
+	if !ok {
+		return "", werror.ErrorWithContextParams(ctx, "executor does not have a spark app id label, could not create label selector")
+	}
+	applicationPods, err := s.podLister.Pods(executor.Namespace).List(labels.Set(map[string]string{common.SparkAppIDLabel: executorSparkLabel}).AsSelector())
+	applicationPodNames := make([]string, 0, len(applicationPods))
+	for _, pod := range applicationPods {
+		applicationPodNames = append(applicationPodNames, pod.Name)
+	}
+	svc1log.FromContext(ctx).Info("Found existing application pods", svc1log.SafeParam("applicationPodNames", applicationPodNames))
+	if err != nil {
+		return "", err
+	}
+
+	// Filter to scheduled pods
+	scheduledPods := make([]*v1.Pod, 0, len(applicationPods))
+	for _, pod := range applicationPods {
+		if pod.Status.Phase == v1.PodPending {
```

Comment:
Change phase check to == `Running`. This ensures we dont try to check for completed pods